textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -293,6 +294,133 @@ private static String toLuceneQueryString(GramBooleanQuery query) {
 			}
 		}
 	}
+	
+	public boolean isEmpty() {
+		if (this.operandSet.size() > 0) {
+			return false;
+		}
+		for (GramBooleanQuery subQuery : this.subQuerySet) {
+			if (! subQuery.isEmpty()) {
+				return false;
+			}
+		}
+		return true;
+	}
+	
+	// "AND" two DNF tree
+	// Apply distributive laws: 
+	// (a OR b) AND (c OR d) --> (a AND c) OR (a AND d) OR (b AND c) OR (c AND d)
+	private static GramBooleanQuery andDNF(GramBooleanQuery left, GramBooleanQuery right) {
```

Comment:
This function is basically: for each node in left query, AND it with each node in right query. 
It's long because we put "leaf nodes" in a list, and "nodes with children" in another list. If we treat leaf nodes and non-leaf nodes as the same in the first place, this function will be much more simpler.